### PR TITLE
Initial charm

### PR DESCRIPTION
**Pull request by weiiwang01**
### Overview

Create the initial `cloudflared` charm.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -35,83 +28,89 @@ def __init__(self, *args: typing.Any):
...
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
-        """Define and start a workload using the Pebble API.
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.config_changed, self._reconcile)
+        self.framework.observe(self.on.secret_changed, self._reconcile)
+        self.framework.observe(self.on["cloudflared-route"].relation_changed, self._reconcile)
+        self._cloudflared_route = CloudflaredRouteRequirer(self)
+        try:
+            tunnel_tokens = self._get_tunnel_tokens()
+        except ValueError:
```
Comment (gregory-schiano) [inline src/charm.py:38]: We should ensure no other exceptions can't be thrown by this method to avoid having the charm in ErrorStatus in the constructor

File: src/charm.py [hunk 2]
```diff
@@ -35,83 +28,89 @@ def __init__(self, *args: typing.Any):
...
+            )
+            return
+        self._config_cloudflared_snap({"tokens": ",".join(tunnel_tokens)})
+        self.unit.status = ops.ActiveStatus()
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+    def _install_cloudflared_snap(self) -> None:  # pragma: nocover
+        """Install the charmed-cloudflared snap."""
+        snap.install_local("./src/charmed-cloudflared_2024.9.1_amd64.snap.zip", dangerous=True)
```
Comment (gregory-schiano) [inline src/charm.py:67]: I guess that's until the snap gets reviewed ?

**Comment by jdkandersson** [comment]
Could the access of this repo please be updated? `is-devops-leadership` should be admin and `is-charms` maintainer

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -5,35 +5,85 @@
...
+    await model.integrate(base_charm.name, cloudflared_charm.name)
+    tunnel_token = cloudflare_api.create_tunnel_token()
+    _, secret_id, _ = await ops_test.juju(
+        "add-secret", "test-tunnel-token", f"tunnel-token={tunnel_token}"
+    )
+    secret_id = secret_id.strip()
+    await model.grant_secret("test-tunnel-token", cloudflared_charm.name)
+    await cloudflared_charm.set_config({"tunnel-token": secret_id})
+    await model.wait_for_idle()
+    wait_for_tunnel_healthy(cloudflare_api, tunnel_token)
```
Comment (jdkandersson) [inline tests/integration/test_charm.py:54]: Is there anything that we could do to check traffic is flowing? Or perhaps that would be too complicated? That would just make sure we have everything working end to end as intended

**Review by jdkandersson — COMMENTED** [review]
I think this broadly looks good, thanks for putting this together quickly!

File: lib/charms/cloudflare_configurator/v0/cloudflared_route.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+r"""# Interface Library for ingress.
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply lib/charms/cloudflare_configurator/v0/cloudflared_route.py]: Fixed, thanks!

File: src/charm.py (hunk 1 above)
Comment (weiiwang01) [reply src/charm.py:38]: If any other exception is thrown at this step, the charm should enter the error state, as it is an unknown error in that situation.

File: src/charm.py (hunk 2 above)
Comment (weiiwang01) [reply src/charm.py:67]: Yes, until the snap is published

File: src/charmed-cloudflared_2024.9.1_amd64.snap.zip
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charmed-cloudflared_2024.9.1_amd64.snap.zip:1]: Yes, until the snap is published

File: tests/integration/conftest.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    """Cloudflare API."""
+
+    def __init__(self, account_id, api_token) -> None:
+        """Initialize the Cloudflare API.
+
+        Args:
+            account_id: cloudflare account ID.
+            api_token: cloudflare API token.
+        """
+        self._endpoint = f"https://api.cloudflare.com/client/v4/accounts/{account_id}/cfd_tunnel"
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply tests/integration/conftest.py:41]: Yes, I will change the Github secrets later

File: tests/integration/test_charm.py (hunk 3 above)
Comment (weiiwang01) [reply tests/integration/test_charm.py:54]: Unless the test account has purchased domains, which requires payment to be added to the account.

File: src/charm.py [hunk 6]
```diff
@@ -35,83 +28,89 @@ def __init__(self, *args: typing.Any):
...
-                    },
-                }
-            },
-        }
+        return self._cloudflared_route.get_tunnel_tokens()
 
 
 if __name__ == "__main__":  # pragma: nocover
-    ops.main.main(IsCharmsTemplateCharm)
+    ops.main.main(CloudflaredCharm)
```
Comment (gregory-schiano) [inline src/charm.py:116]: Should be `ops.main` as of 2.16
https://ops.readthedocs.io/en/latest/#ops-main-entry-point

Comment (weiiwang01) [reply src/charm.py]: Updated, thanks!

File: src/charm.py (hunk 1 above)
Comment (jdkandersson) [reply src/charm.py:38]: Let's just count the number of relations

File: src/charm.py [hunk 7]
```diff
@@ -32,85 +41,143 @@ def __init__(self, *args: typing.Any):
...
+                "--dangerous",
+                "./src/charmed-cloudflared_2024.9.1_amd64.snap.zip",
+            ]
+        )
 
-    def _config_cloudflared_snap(self, config: dict[str, str]) -> None:  # pragma: nocover
+    @staticmethod
+    def _config_cloudflared_snap(
+        name: str, config: dict[str, str | int]
+    ) -> None:  # pragma: nocover
```
Comment (gregory-schiano) [inline src/charm.py:111]: Add a comment to justify why you need to ignore coverage

File: src/charm.py [hunk 8]
```diff
@@ -32,85 +41,143 @@ def __init__(self, *args: typing.Any):
...
-        """Receive tunnel tokens from all configuration sources.
+        charmed_cloudflared = snap.SnapCache()[name]
+        if all(charmed_cloudflared.get(key) == str(value) for key, value in config.items()):
+            return
+        charmed_cloudflared.set(config, typed=True)
 
-        Returns:
-            Cloudflared tunnel tokens.
+    @staticmethod
+    def _remove_cloudflared_snap(name: str) -> None:  # pragma: nocover
```
Comment (gregory-schiano) [inline src/charm.py:124]: Add a comment to justify why you need to ignore coverage

File: src/charm.py
Comment (gregory-schiano) [inline src/charm.py:1]: Instead of using "@staticmethod" to simplify unittesting, I'd suggest to not cover these methods with unit tests but with integration tests, keep ignoring coverage for these methods but add a comment to justify. And most important, ensure a good (happy / edge cases) integration test coverage.
Therefore you can move the methods in a dedicated module, or at least put them at module level and remove the decorator

File: lib/charms/cloudflare_configurator/v0/cloudflared_route.py [hunk 9]
```diff
@@ -100,33 +105,18 @@ def __init__(
...
-            secret = self._charm.model.get_secret(id=secret_id)
-            tunnel_token = secret.get_content(refresh=True)[_TUNNEL_TOKEN_SECRET_VALUE_FIELD]
-            tunnel_tokens.append(tunnel_token)
-        return tunnel_tokens
+        relation_data = relation.data[relation.app]
+        secret_id = relation_data.get(_TUNNEL_TOKEN_SECRET_ID_FIELD)
+        if not secret_id:
+            return None
+        secret = self._charm.model.get_secret(id=secret_id)
+        return secret.get_content(refresh=True)[_TUNNEL_TOKEN_SECRET_VALUE_FIELD]
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply lib/charms/cloudflare_configurator/v0/cloudflared_route.py]: The charm now went to blocked state in this case, thanks!

File: src/charm.py [hunk 10]
```diff
@@ -32,85 +41,143 @@ def __init__(self, *args: typing.Any):
...
-                for metrics_port in range(15300, 15300 + len(tunnel_tokens))
+                for metrics_port in self._get_instance_metrics_ports().values()
             ],
             dashboard_dirs=["./src/grafana_dashboards"],
         )
 
     def _on_install(self, _: ops.EventBase) -> None:
         """Install the charmed-cloudflared snap."""
-        self._install_cloudflared_snap()
+        # pylint: disable=protected-access
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charm.py:62]: Added a link to the snap parallel installs document

File: src/charm.py [hunk 11]
```diff
@@ -32,85 +41,143 @@ def __init__(self, *args: typing.Any):
...
-        config_tunnel_tokens = self._get_tunnel_tokens_from_config()
-        integration_tunnel_tokens = self._get_tunnel_token_from_integration()
-        if config_tunnel_tokens and integration_tunnel_tokens:
-            raise ValueError("received tunnel-tokens from config and integration")
-        return config_tunnel_tokens or integration_tunnel_tokens
+        snap.remove(name)
 
-    def _get_tunnel_tokens_from_config(self) -> list[str]:
-        """Receive tunnel tokens from charm configuration.
+    def _installed_cloudflared_snaps(self) -> set[str]:  # pragma: nocover
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charm.py]: This method has been removed.

File: src/charm.py [hunk 12]
```diff
@@ -32,85 +41,143 @@ def __init__(self, *args: typing.Any):
...
-        tokens = []
-        secret_id = typing.cast(str, self.config.get("tunnel-token"))
-        if secret_id:
-            secret = self.model.get_secret(id=secret_id)
-            tokens.append(secret.get_content(refresh=True)["tunnel-token"])
-        return tokens
+        installed_charmed_cloudflared = set()
+        installed_snaps = self._snap_client.get_installed_snaps()
+        for installed_snap in installed_snaps:
+            if installed_snap["name"].startswith("charmed-cloudflared"):
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charm.py]: Updated to use a constant

File: src/charm.py [hunk 13]
```diff
@@ -32,85 +41,143 @@ def __init__(self, *args: typing.Any):
...
+            A mapping of charmed-cloudflared snap instance name to metrics ports.
         """
-        return self._cloudflared_route.get_tunnel_tokens()
+        metrics_ports = {}
+        if self.config.get(TUNNEL_TOKEN_CONFIG_NAME):
+            metrics_ports["charmed-cloudflared_config0"] = 15299
+        for relation in self.model.relations[CLOUDFLARED_ROUTE_INTEGRATION_NAME]:
+            if relation.app is None:
+                continue
+            metrics_ports[f"charmed-cloudflared_relation{relation.id}"] = 15300 + relation.id
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charm.py]: Updated the naming.

File: src/charm.py [hunk 14]
```diff
@@ -32,85 +41,143 @@ def __init__(self, *args: typing.Any):
...
 
     def _reconcile(self, _: ops.EventBase) -> None:
         """Handle changed configuration."""
         try:
-            tunnel_tokens = self._get_tunnel_tokens()
-        except ValueError:
-            self.unit.status = ops.BlockedStatus(
-                "tunnel-token is provided in both the config and integration"
-            )
+            metrics_ports = self._get_instance_metrics_ports()
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charm.py:68]: Updated, thanks!

File: src/charm.py (hunk 7 above)
Comment (weiiwang01) [reply src/charm.py]: This method has been removed.

File: src/charm.py (hunk 8 above)
Comment (weiiwang01) [reply src/charm.py]: This method has been removed.

File: src/charm.py
Comment (weiiwang01) [reply src/charm.py:1]: Yes, updated, thanks!

File: src/charm.py [hunk 15]
```diff
@@ -35,83 +41,122 @@ def __init__(self, *args: typing.Any):
...
+        if tunnel_token_config:
+            try:
+                secret = self.model.get_secret(id=tunnel_token_config)
+                secret_value = secret.get_content(refresh=True)["tunnel-token"]
+                tunnel_tokens[f"{CHARMED_CLOUDFLARED_SNAP_NAME}_config0"] = secret_value
+            except (ops.SecretNotFoundError, ops.ModelError, KeyError) as exc:
+                raise InvalidConfig("invalid tunnel-token config") from exc
+        relations = self.model.relations[CLOUDFLARED_ROUTE_INTEGRATION_NAME]
+        if tunnel_tokens and relations:
+            raise InvalidConfig("tunnel-token is provided by both the config and integration")
```
Comment (gregory-schiano) [inline src/charm.py:128]: Can't you do the check at the beg on the method based on the presence of `tunnel_token_config` ?
That'll avoid the `if tunnel_token_config`
I know there might be the case where the tunnel config is pointing to a non existing secret, but feature wise, if you set the config you'd expect the charm to use it and not the relation. So I'd rather rely on the config option being set

File: src/charm.py [hunk 16]
```diff
@@ -66,75 +71,38 @@ def _reconcile(self, _: ops.EventBase) -> None:
...
+        for instance, tunnel_token in tunnel_tokens.items():
+            charmed_cloudflared = snap.SnapCache()[instance]
+            config = {
+                "tunnel-token": tunnel_token,
+                "metrics-port": metrics_ports[instance],
+            }
+            if all(charmed_cloudflared.get(key) == str(value) for key, value in config.items()):
+                continue
+            logger.info("configuring charmed-cloudflared instance: %s", instance)
+            charmed_cloudflared.set(config, typed=True)
```
[the remark this answers is not in the document]
Comment (weiiwang01) [reply src/charm.py:104]: The snap will automatically restart the tunnel process once the snap configuration is changed (in the `configure` snap hook).

File: src/charm.py (hunk 15 above)
Comment (weiiwang01) [reply src/charm.py]: Yes, I moved the check to the beginning of this method

File: .github/workflows/integration_test.yaml [hunk 17]
```diff
@@ -5,18 +5,9 @@ on:
 
 jobs:
   integration-tests:
-    uses: canonical/operator-workflows/.github/workflows/integration_test.yaml@main
```
Comment (amandahla) [inline .github/workflows/integration_test.yaml:8]: secret-env?

File: src/charm.py [hunk 18]
```diff
@@ -35,83 +41,122 @@ def __init__(self, *args: typing.Any):
...
+        relations = self.model.relations[CLOUDFLARED_ROUTE_INTEGRATION_NAME]
+        if tunnel_token_config and relations:
+            raise InvalidConfig("tunnel-token is provided by both the config and integration")
+        if tunnel_token_config:
+            try:
+                secret = self.model.get_secret(id=tunnel_token_config)
+                secret_value = secret.get_content(refresh=True)["tunnel-token"]
+                return {f"{CHARMED_CLOUDFLARED_SNAP_NAME}_config0": secret_value}
+            except (ops.SecretNotFoundError, ops.ModelError, KeyError) as exc:
+                raise InvalidConfig("invalid tunnel-token config") from exc
```
Comment (amandahla) [inline src/charm.py:127]: This is the message that will appear in BlockedStatus, right? In this case could you provide more info that it was raised while getting the secret?

**Review by gregory-schiano — DISMISSED** [review]

File: .github/workflows/integration_test.yaml (hunk 17 above)
Comment (weiiwang01) [reply .github/workflows/integration_test.yaml:8]: Yes, this uses a new feature that has not yet been merged into the `operator-workflows` to provide secrets for the integration tests. The new method should offer better maintainability and be more robust against leaks.

https://github.com/canonical/operator-workflows/pull/427
